File: AoC/2023_problems/day3.py
from collections import defaultdict

test = """467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."""

with open("AoC/2023_problems/day3.txt") as f:
    test = f.read()

rows = [list(x) for x in test.split()]

def check_neighbours(locations):
    s = ".0123456789"
    if len(locations) > 2:
        locations = [locations[0], locations[2]]
    neighbours = []
    gear = set()
    for loc in locations:
        i, j = loc
        for m in range(i-1, i+2):
            for n in range(j-1, j+2):
                if (m < 0 or m >= len(rows)) or (n < 0 or n >= len(rows[0])):
                    continue
                if rows[m][n] not in s:
                    neighbours.append(True)
                    if rows[m][n] == "*":
                        gear.add((m, n))
    return any(neighbours), gear

total = 0
gears = defaultdict(list)
for i in range(len(rows)):
    num = ""
    locations = []
    for j in range(len(rows[0])):
        if rows[i][j].isdigit():
            num += rows[i][j]
            locations.append([i, j])
        elif num:
            add, gear = check_neighbours(locations)
            if add:
                total += int(num)
            if gear:
                for g in gear:
                    gears[g].append(int(num))
            num = ""
            locations = []
    # End of row check
    if num:
        add, gear = check_neighbours(locations)
        if gear:
            for g in gear:
                gears[g].append(int(num))
        if add:
            total += int(num)
print(total)
total_gears = 0
for gear in gears:
    if len(gears[gear]) > 1:
        total_gears += gears[gear][0] * gears[gear][1]
print(total_gears)

File: AoC/2023_problems/day3.js
const fs = require("fs");

const data = fs.readFileSync("AoC/2023_problems/day3.txt", "utf8");
const dataRows = data.split("\n");
// Pad the table around with dots for easier parsing later on
const pad = Array(dataRows[0].length + 2).fill(".");
let rows = [pad];
for (const row of dataRows) {
    rows.push([".", ...row, "."]);
}
rows.push(pad);

const ignoreList = ".0123456789";
const isDigit = (x) => ignoreList.slice(1).includes(x);
const isSymbol = (x) => !ignoreList.includes(x);

// Checks neighboring cells for symbols and gears and returns true if a symbol is found, false otherwise.
// Adds new gears to the set in the form of a string "i,j" coordinates
function checkNeighbors(row, col, gearsCoord) {
    let symbolFound = false;
    let neighbors = [
        [-1, -1], [0, -1], [1, -1],
        [-1, 0], [1, 0],
        [-1, 1], [0, 1], [1, 1] , 
    ];
    for (let [x, y] of neighbors) {
        if (isSymbol(rows[row + x][col + y])) {
            symbolFound = true;
        }
        if (rows[row + x][col + y] === "*") {
            gearsCoord.add(`${row + x},${col + y}`);
        }
    }
    return symbolFound;
}

function solve(rows) {
    let nums = [];
    let gears = {};
    for (let i = 1; i < rows.length - 1; i++) {
        let j = 1;
        while (j < rows[0].length) {
            if (isDigit(rows[i][j])) {
                let symbolChecks = [];
                let gearCoord = new Set();
                let num = "";
                // Collect all numbers that are next to each other
                while (isDigit(rows[i][j])) {
                    num += rows[i][j];
                    // Adding new gears to the set and checking for symbols next to the current i,j cell
                    symbolChecks.push(checkNeighbors(i, j, gearCoord));
                    j++;
                }
                // Check if any symbols were found and add num to the nums
                if (symbolChecks.some(Boolean)) {
                    nums.push(parseInt(num));
                }
                // Add collected gears to the set. If gear already exists, push to the array.
                for (let gc of gearCoord) {
                    if (gears.hasOwnProperty(gc)) {
                        gears[gc].push(parseInt(num));
                    }
                    else {
                        gears[gc] = [parseInt(num)];
                    }
                }
            }
            j++;
        }
    }
    let p1 = nums.reduce((acc,x) => acc + x, 0);
    let p2 = 0;
    for (let [k, v] of Object.entries(gears)) {
        if (v.length > 1) {
            p2 += v[0] * v[1];
        }
    }
    return [p1, p2];
}

console.log(solve(rows));
